Crash when sharing images to AnkiDroid

Steps:

[Screen_recording_20240224_200611.webm](https://github.com/ankidroid/Anki-Android/assets/69634269/a8ef5b7c-e4e9-4833-9559-995c8f0106d3)


Debug info
```text
AnkiDroid Version = 2.18alpha1-debug (f145def4c4ee9b944bcb1df6cd04a636a5cc8bd3)

Backend Version = 0.1.34-anki23.12.1 (23.12.1 1a1d4d5419c6b57ef3baf99c9d2d9cf85d36ae0a)

Android Version = 12 (SDK 31)

ProductFlavor = full

Manufacturer = Google

Model = sdk_gphone64_x86_64

Hardware = ranchu

Webview User Agent = Mozilla/5.0 (Linux; Android 12; sdk_gphone64_x86_64 Build/SE1A.211212.001.B1; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/121.0.6167.164 Mobile Safari/537.36

ACRA UUID = 1135e961-f358-44c5-aa32-5f3288eeaf72

Crash Reports Enabled = false
```

```text
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ichi2.anki.debug/com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity}: java.lang.NumberFormatException: For input string: "raw:/storage/emulated/0/Download/pexels-pixabay-36717.jpg"
         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3635)
         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792)
         at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
         at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
         at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
         at android.os.Handler.dispatchMessage(Handler.java:106)
         at android.os.Looper.loopOnce(Looper.java:201)
         at android.os.Looper.loop(Looper.java:288)
         at android.app.ActivityThread.main(ActivityThread.java:7839)
         at java.lang.reflect.Method.invoke(Native Method)
         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
        Caused by: java.lang.NumberFormatException: For input string: "raw:/storage/emulated/0/Download/pexels-pixabay-36717.jpg"
         at java.lang.Long.parseLong(Long.java:594)
         at java.lang.Long.parseLong(Long.java:636)
         at com.ichi2.anki.multimediacard.fields.BasicImageFieldController.getImageNameFromUri(BasicImageFieldController.kt:787)
         at com.ichi2.anki.multimediacard.fields.BasicImageFieldController.internalizeUri(BasicImageFieldController.kt:526)
         at com.ichi2.anki.multimediacard.fields.BasicImageFieldController.directImageIntent(BasicImageFieldController.kt:486)
         at com.ichi2.anki.multimediacard.fields.BasicImageFieldController.createUI(BasicImageFieldController.kt:226)
         at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.recreateEditingUi(MultimediaEditFieldActivity.kt:191)
         at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.recreateEditingUi$default(MultimediaEditFieldActivity.kt:167)
         at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.intentEditImage(MultimediaEditFieldActivity.kt:118)
         at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.onCreate(MultimediaEditFieldActivity.kt:107)
         at android.app.Activity.performCreate(Activity.java:8051)
         at android.app.Activity.performCreate(Activity.java:8031)
         at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608)
         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792) 
         at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
         at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
         at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210) 
         at android.os.Handler.dispatchMessage(Handler.java:106) 
         at android.os.Looper.loopOnce(Looper.java:201) 
         at android.os.Looper.loop(Looper.java:288) 
         at android.app.ActivityThread.main(ActivityThread.java:7839) 
         at java.lang.reflect.Method.invoke(Native Method) 
         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003) 
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.kt
@@ -774,7 +774,7 @@ class BasicImageFieldController : FieldControllerBase(), IFieldController {
      *
      * @return Display name of file identified by uri (null if does not exist)
      */
-    private fun getImageNameFromUri(context: Context, uri: Uri): String? {
+    private fun getImageNameFromUri(context: Context, uri: Uri): String? = try {
         Timber.d("getImageNameFromUri() URI: %s", uri)
         var imageName: String? = null
         if (DocumentsContract.isDocumentUri(context, uri)) {
@@ -784,19 +784,44 @@ class BasicImageFieldController : FieldControllerBase(), IFieldController {
                 val selection = MediaStore.Images.Media._ID + "=" + id
                 imageName = getImageNameFromContentResolver(context, MediaStore.Images.Media.EXTERNAL_CONTENT_URI, selection)
             } else if ("com.android.providers.downloads.documents" == uri.authority) {
-                val contentUri = ContentUris.withAppendedId(Uri.parse("content://downloads/public_downloads"), docId.toLong())
-                imageName = getImageNameFromContentResolver(context, contentUri, null)
+                imageName = when {
+                    // `msf:1000034860` can be handled by using the URI
+                    docId.startsWith(MEDIASTORE_DOWNLOAD_FILE_PREFIX) -> {
+                        getImageNameFromContentResolver(context, uri, null)
+                    }
+
+                    // raw:/storage/emulated/0/Download/pexels-pixabay-36717.jpg
+                    docId.startsWith(RAW_DOCUMENTS_FILE_PREFIX) -> {
+                        docId.substring(RAW_DOCUMENTS_FILE_PREFIX.length).split("/").toTypedArray().last()
+                    }
+
+                    docId.toLongOrNull() != null -> {
+                        val contentUri = ContentUris.withAppendedId(Uri.parse("content://downloads/public_downloads"), docId.toLong())
+                        getImageNameFromContentResolver(context, contentUri, null)
+                    }
+
+                    else -> {
+                        CrashReportService.sendExceptionReport(
+                            message = "Failed to get fileName from providers.downloads.documents",
+                            origin = "getImageNameFromUri"
+                        )
+                        null
+                    }
+                }
             }
         } else if ("content".equals(uri.scheme, ignoreCase = true)) {
             imageName = getImageNameFromContentResolver(context, uri, null)
         } else if ("file".equals(uri.scheme, ignoreCase = true)) {
             if (uri.path != null) {
-                val pathParts = uri.path!!.split("/").toTypedArray()
-                imageName = pathParts[pathParts.size - 1]
+                imageName = uri.path!!.split("/").last()
             }
         }
         Timber.d("getImageNameFromUri() returning name %s", imageName)
-        return imageName
+        imageName
+    } catch (e: Exception) {
+        Timber.w(e)
+        CrashReportService.sendExceptionReport(e, "getImageNameFromUri")
+        null
     }
 
     /**
@@ -888,6 +913,18 @@ class BasicImageFieldController : FieldControllerBase(), IFieldController {
         private const val TAKE_PICTURE_LAUNCHER_KEY = "take_picture_launcher_key"
         private const val SELECT_IMAGE_LAUNCHER_KEY = "select_image_launcher_key"
         private const val DRAWING_LAUNCHER_KEY = "drawing_launcher_key"
+
+        /**
+         * https://cs.android.com/android/platform/superproject/+/master:packages/providers/DownloadProvider/src/com/android/providers/downloads/MediaStoreDownloadsHelper.java;l=24
+         */
+        private const val MEDIASTORE_DOWNLOAD_FILE_PREFIX = "msf:"
+
+        /**
+         * The default prefix to raw file documentIds
+         *
+         * https://cs.android.com/android/platform/superproject/+/master:packages/providers/DownloadProvider/src/com/android/providers/downloads/RawDocumentsHelper.java;l=35?q=%5C%22raw:%5C%22&ss=android%2Fplatform%2Fsuperproject
+         */
+        const val RAW_DOCUMENTS_FILE_PREFIX = "raw:"
 
         /**
          * Get Uri based on current image path

